contents/linux/shell.md

```diff
@@ -0,0 +1,180 @@
+
+- **Remote Management**: Use over remote command-line shell makes it much easier to perform tasks remotely, such as when a developer is connecting from home. A remote command-line shell usually requires less bandwidth.
+
+- **Stability**: No matter what linux distribution you are running, the command line shell is always available and usage is almost the same. However, GUI tools might be different in different distributions, make it very hard to learn.
+
+- **Freedom**: You will have complete control in command-line shell. GUIs usually tends to simplifying things, giving user fewer options to make things easier. In command line there is no such simplifications, you will have complete control of what you are doing.
+
+- **Simplicity**: Although you might find command-line hard to use at first, but after using it enough you'll eventually come to the point where the command-line is far easier to use than trying to find options in the drop-down menus and alike of a GUI. 
+
+### Benefit: Program with bash script
+
+One important feature in bash is **bash script**. It enables us to do `task automation`. Suppose you are a system admin, there is a task you need to perform daily without any changes. Instead of typing in the set of commands for the task everyday, you can store these commands in a file and run the script.
+
+Here's one example of bash script, suppose you have to check if some given file exist periodically. The following script will ask for your input and output the results. `filename` is a variable used to store user input, `filetype` and `perm` stores corresponding attributes of the file for output.
+
+```bash
+#!/bin/bash
+ 
+echo -e "Please input a filename, I will check the filename's type and permission.\n\n"
+
+# waiting for user input.
+read -p "Input a filename : "  filename
+
+# check if file exists.
+test -z $filename && echo "You MUST input a filename." && exit 0
+test ! -e $filename && echo "The filename '$filename' DO NOT EXIST" && exit 0
+
+# check file type.
+test -f $filename && filetype="regular file"
+test -d $filename && filetype="directory"
+
+# check file permission.
+test -r $filename && perm="readable"
+test -w $filename && perm="$perm writable"
+test -x $filename && perm="$perm executable"
+
+# output result.
+echo "The filename: $filename is a $filetype"
+echo "And permissions are : $perm"
+```
+
+The above bash script is still just a collection of bash commands. Although it saves us a lot of works if it is a daily routine task, it still can be done by typing those commands line by line. What makes bash script really exciting is the semantics of `conditionals`, `loops` and `functions`. Which makes bash script [**turning complete**](https://en.wikipedia.org/wiki/Turing_completeness).
```

Comment:
Grammar. Maybe bold bash script?
```suggestion
The above bash script is still just a collection of bash commands. Although it saves us a lot of work if it is a routine task, it can still be done by typing those commands line by line. What makes bash script really exciting is the semantics of `conditionals`, `loops` and `functions`. Which makes bash script [**turning complete**](https://en.wikipedia.org/wiki/Turing_completeness).
```

Reply:
Why do we care that Bash scripts are Turing complete? I feel that this is a theoretical point that may not be of practical relevance; the above example is good enough to demonstrate its power.